File: src/dataStructures/stack.rb
class Node
    attr_accessor :data, :next

    def initialize(data)
        @data = data
        @next = nil
    end
end

class Stack
    attr_accessor :first, :last, :size

    def initialize
        @first = nil
        @last = nil
        @size = 0
    end


    def push(data)
        new_node = Node.new(data)
        if !@first
            @first = new_node
            @last = new_node
        else 
            old_head = @first
            @first = new_node
            new_node.next = old_head
        end
        return @size += 1
    end

    def pop
        return nil if @size==0
        old_head = @first
        if @first==@last
            @last = nil
        end
        @first = @first.next
        @size -= 1
        return old_head
    end

    def is_empty?
        return !!@first
    end

end

File: src/dataStructures/stack.js
class Node {
    constructor(val){
    this.val = val
    this.next = null
    }
}

class Stack {   // using a Singly Linked List with push() adding to the beginning and pop() removing from the beginning
    constructor() {
        this.first = null
        this.last = null
        this.size = 0
    }

    push(val) { // adds a Node to the BEGINNING of a Stack in constant time - Big O(1)
        let newNode = new Node(val)
        if (this.size===0) {
            this.first = newNode
            this.last = newNode
        } else {
            let oldHead = this.first
            this.first = newNode
            newNode.next = oldHead
        }
        return ++this.size
    }

}